### DISCUSSION TITLE
ESM-only oidc-provider and NestJS

### DISCUSSION THREAD

**Discussion by Gamote** [panva/node-oidc-provider#1249] [Community Guides]
First of all, amazing package! 💯  Huge thanks to all the contributors for your time put into this project. 🙏🏼 

## What is up?
As you all know since `v8.x` this package is [delivered as ESM only](https://github.com/panva/node-oidc-provider/releases/tag/v8.0.0#:~:text=tokens%20(cb67083)-,oidc%2Dprovider%20is%20now%20an%20ESM%2Donly%20module,-(3c5ebe1)). NestJS being a CommonJS only (at least for the foreseeable future). 
I assume that there is no option for the package to also deliver CommonJS (`?`) so I had to find a solution. Solution that I want to share with you, hopping that it will help developers and/or maybe it can be improved.

---

### Summary of the issue
- Nest doesn't support ESM ([issue](https://github.com/nestjs/nest/issues/7021))
- `oidc-provider` is a full ESM package since version 8 ([release](https://github.com/panva/node-oidc-provider/releases/tag/v8.0.0))
- We don't want to migrate Nest to ESM, until it's officially supported

Read more about it:
- https://github.com/microsoft/TypeScript/issues/43329
- https://github.com/nestjs/nest/issues/7021

### Solution

The solution was to have the module asynchronously be imported. Then we created an injectable provider that will await the import of the `oidc-provider`. This is so we don't have be forced to await it down the line. For example if we want to use it in a constructor. Code:

```ts
// oidc-provider-module.ts
import { FactoryProvider } from '@nestjs/common';
// @ts-ignore
import type ESMProvider, { interactionPolicy, errors } from 'oidc-provider';

/**
 * We re-export the type so the consumer doesn't have to import it from the oidc-provider package and be forced to ignore the type error.
 */
export type Provider = ESMProvider;

/**
 * A type representing the actual provider class provided by the `oidc-provider` package and not the instance.
 */
export type ProviderClass = { new (...args: ConstructorParameters<typeof ESMProvider>): Provider };

/**
 * A type representing the `oidc-provider` module.
 * Useful when we pass the around as a parameter.
 */
export type OidcProviderModule = { Provider: ProviderClass, interactionPolicy: typeof interactionPolicy, errors: typeof errors };

/**
 * Dynamically imports a package. Useful when we want to import an ES module into CommonJS.
 *
 * **Note:** We couldn't use only `await import(packageName)` because it still throws the ESM error.
 * @see https://github.com/microsoft/TypeScript/issues/43329#issuecomment-811606238
 *
 * @param packageName
 */
const dynamicImport = async <ReturnType>(packageName: string): Promise<ReturnType> =>
  new Function(`return import('${packageName}')`)();

/**
 * The OIDC provider module token. It's used to inject the OIDC provider as a dependency like so:
 * ```ts
 * @Inject(OIDC_PROVIDER_MODULE) private readonly oidcProviderModule: OidcProviderModule
 * ```
 */
export const OIDC_PROVIDER_MODULE = 'OIDC_PROVIDER_MODULE';

/**
 * A factory provider that provides the OIDC provider module.
 */
export const oidcProviderModuleFactory = (): FactoryProvider => ({
  provide: OIDC_PROVIDER_MODULE,
  useFactory: async (): Promise<OidcProviderModule> => {
    const { default: ProviderClass, ...rest } = await dynamicImport<{ default: ProviderClass } & Omit<OidcProviderModule, 'Provider'>>(
      'oidc-provider',
    );

    return {
      Provider: ProviderClass,
      ...rest,
    };
  },
});
```

> Ideally this whole wrapper is removed and we use the package directly.

### Usage

1. Add the provider in the module
    
    ```ts
    // oidc.module.ts
    import { oidcProviderModuleFactory } from './oidc-provider-module';
    
    @Module({
      providers: [oidcProviderModuleFactory(), OidcService],
    })
    export class OidcModule {}
    ```

2. Use it in the service
    
    ```ts
    // oidc.service.ts
    import { OIDC_PROVIDER_MODULE, OidcProviderModule, Provider } from './oidc-provider-module';
    
    @Injectable()
    export class OidcService {
      private readonly oidc: Provider;
    
      constructor(@Inject(OIDC_PROVIDER_MODULE) private readonly opm: OidcProviderModule) {
        this.oidc = new opm.Provider('http://localhost:3000', {
          // the rest of your configuration...
        });
      }
    }
    ```

**Comment by natnt2000** [comment]
@Gamote how can I set it up with ExpressJS?
My project is using commonjs

**Reply by Gamote** [reply]
Convert the Express project if possible,or use a bundler. If that is not possible, then create a similar functionality that "initialise" the library (a singleton class can be used for example) and delay the "listen" and any other dependent logic until the module is initialised.

**Comment by panva** [comment]
> I assume that there is no option for the package to also deliver CommonJS (?) so I had to find a solution.

There's no going back to CJS for this module.

That being said, be on a lookout for https://github.com/nodejs/node/pull/51977 to land in the supported Node.js release lines. With `--experimental-require-module`, and later on when it goes out of experimental, you'll be able to `require(esm)` so long as the module itself doesn't use any top level await, which I don't ever plan on doing with my modules.

Refs: https://joyeecheung.github.io/blog/2024/03/18/require-esm-in-node-js/

**Reply by JonasDoe** [reply]
Well, even with that flag the default export gave me a hard time. I'm sure there are reasons for this decision, but it's tough. I figure we're gonna rewrite our project resp. start over from the scratch with ESM and see which other dependencies we can still use.

**Reply by panva** [reply]
🤷 

> ➜  node -v
> v22.0.0
> ➜  node --experimental-require-module -p "const { default: Provider } = require('oidc-provider'); Provider"
> [class Provider extends EventEmitter]

**Reply by JonasDoe** [reply]
Well, I figure I should have been more specific - the JavaScript is generated by TypeScript, which will emit code like
```javascript
const oidc_provider_1 = require("oidc-provider");
...
new oidc_provider_1.Provider(...);
```
instead of
```javascript
const { default: Provider } = require('oidc-provider');
...
new Provider(...);
```
Gotta confess, I'm not really into that matter, maybe I'm missing the obvious. For me it seems like restarting with ESM is the most substantial approach. :/

**Reply by panva** [reply]
If you know you transpile ESM to CJS you can just 

```
import { default as Provider } from 'oidc-provider';
```

This ought to get you the CJS you need for that flag and default export to kick in.

**Reply by JonasDoe** [reply]
Sadly not, it transpiles to:
```javascript
const oidc_provider_2 = tslib_1.__importDefault(require("oidc-provider"));
...
new oidc_provider_2.default.Provider(...)
```
and the error `oidc_provider_2.default.Provider is not a constructor`.

But it feels close enough, I've edited the generated file to log `oidc_provider_2`, and it reads like:
```
{
  default: [Module: null prototype] {
    default: [class Provider extends EventEmitter],
    errors: [Module: null prototype] {
      AccessDenied: [class AccessDenied extends OIDCProviderError],
...
```
`oidc_provider_2.default.default.Provider(...)` would fail with the same error, despite being the `[class Provider extends EventEmitter]`.

I'm wondering, wouldn't it be possible that the `Provider` gets exported in both ways, i.e.:
```typescript
export default Provider;
export { errors, interactionPolicy, Provider};
```
?

**Reply by panva** [reply]
> as export default Provider as export Provider?

Yeah, I'll do that for sure if your issue prevails.

> Sadly not, it transpiles to:

That's however not because of how node treats the module, that's something in your transpile step trying to be smart and wrapping the module in another default. I feel like that's something you should be able to turn off in your tsconfig somehow.

**Reply by JonasDoe** [reply]
Hm, I shrunk my `tsconfig.json` until that remained:
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "rootDir": "src",
    "outDir": "dist",
    "target": "ES2020",
    "skipLibCheck": true,
    "esModuleInterop": true
  },
  "exclude": ["node_modules"],
  "include": ["src/**/*"]
}
```
and I'm running `tsc && node --experimental-require-module dist/serve.js`, so there's not much place for fancy magic inbetween. My `tsc` version is `5.4.5` - I could downgrade that one and try again if you feel it is the issue.

But anyway, shouldn't my manual edit to `oidc_provider_2.default.default.Provider(...)` work with that double defaulting? Just wondering whether this wrapping is really the root of all problems, and not just an additional one.

**Reply by panva** [reply]
If you disable esModuleInterop it works as expected. This might as well just be typescript not ready for require(esm) in that it's trying to do something it shouldn't anymore. I suggest you open an issue in typescript.

**Reply by JonasDoe** [reply]
Alright, will do that. Thanks for your patience!

**Reply by JonasDoe** [reply]
Update - seems like the ball is in Node's field: https://github.com/microsoft/TypeScript/issues/58341. So lets hope this experimental feature will make it rather soon!

**Reply by panva** [reply]
Node.js v23 will ship with unflagged require(esm) so oidc-provider will be requireable without any wrappers. This capability will eventually trickle down to LTS Node.js 22.x

**Reply by milo526** [reply]
Unflagged require(esm) has landed in Node.js 22.12.0 (https://nodejs.org/en/blog/release/v22.12.0)

**Reply by panva** [reply]
And will land in [20.19.0 on thursday](https://github.com/nodejs/node/pull/57349). At the same time its support in TypeScript also [landed with 5.8](https://devblogs.microsoft.com/typescript/announcing-typescript-5-8/#support-for-require()-of-ecmascript-modules-in---module-nodenext)

**Comment by ArturChe** [comment]
Hi @Gamote I tried to repeat your solution, but I got an error:
```
Debugger listening on ws://0.0.0.0:9229/e48d22b0-28ce-4aaa-be5c-c69a4e96571e
2024-07-23T05:37:12.161370664Z For help, see: https://nodejs.org/en/docs/inspector
2024-07-23T05:37:13.095780795Z /usr/src/app/js/node_modules/nest-oidc-provider/dist/oidc.service.js:42
2024-07-23T05:37:13.095804925Z const oidc = __importStar(require("oidc-provider"));
2024-07-23T05:37:13.095807145Z                           ^
2024-07-23T05:37:13.095808812Z 
2024-07-23T05:37:13.095810395Z Error [ERR_REQUIRE_ESM]: require() of ES Module /usr/src/app/js/node_modules/oidc-provider/lib/index.js from /usr/src/app/js/node_modules/nest-oidc-provider/dist/oidc.service.js not supported.
2024-07-23T05:37:13.095812300Z Instead change the require of index.js in /usr/src/app/js/node_modules/nest-oidc-provider/dist/oidc.service.js to a dynamic import() which is available in all CommonJS modules.
2024-07-23T05:37:13.095814238Z     at Object.<anonymous> (/usr/src/app/js/node_modules/nest-oidc-provider/dist/oidc.service.js:42:27) {
2024-07-23T05:37:13.095816234Z   code: 'ERR_REQUIRE_ESM'
2024-07-23T05:37:13.095817763Z }
2024-07-23T05:37:13.095819226Z 
2024-07-23T05:37:13.095820657Z 

Node.js v20.11.1
```

**Comment by adrianbrs** [comment]
Hello everyone!

I’d like to share that you can use the [nest-oidc-provider](https://github.com/adrianbrs/nest-oidc-provider) package to integrate `oidc-provider` with Nest. Version [2.0.0](https://github.com/adrianbrs/nest-oidc-provider/releases/tag/2.0.0) now supports the ESM-only version of `oidc-provider` using a similar approach to what was described above.

Hope this helps!

**Reply by panva** [reply]
Node.js v23 will ship with unflagged require(esm) so oidc-provider will be requireable without any wrappers. This capability will eventually trickle down to LTS Node.js 22.x

**Reply by adrianbrs** [reply]
Hi @panva! Yeah, that's great news!

In the Nest ecosystem, it's quite common to use 'wrappers' for integrating with other libraries (Nest itself has many official ones). If you take a closer look at the library, you'll notice it offers more than just ESM support. Both libraries still need to be installed, and `nest-oidc-provider` mainly simplifies configuration and handles dependency injection through decorators, so I believe `nest-oidc-provider` is still a valid and useful option.

Let me know if there's anything I can do to improve this integration, and thanks for creating such an awesome lib!

**Comment by Sufiyan-kryptos** [comment]
Is the issue resolved in express ? we should also support CommonJS. Google firestore doesn't support ESM yet

**Reply by panva** [reply]
CJS style `let oidc = require('oidc-provider')` is possible in Node.js versions where the `require(esm)` feature is enabled by default (^20.19.0 || ^22.12.0 || >= 23.0.0)

**Comment by Sczlog** [comment]
i'm using dynamic import to wrap oidc-provider as an async function to use it in
```
// wrapper.ts
const _oidc = import('oidc-provider');

export const getOidcModule = async () => {
  return await _oidc;
};
// oidc.ts
export const addOidcProvider = async (express: GraphQLServer['express']) => {
  const { Provider, interactionPolicy } = await getOidcModule();
  const provider = new Provider( /**/);
}
```

**Comment by bugerry87** [comment]
### Please Get Rid of Constructor.name 

I hope my issue is in any kind related to this topic of discussion - anyway - we wish to bundle and minify our services.
However, `oidc-provider` does not support minified code since it makes use of `constructor.name` in several lines:

- https://github.com/panva/node-oidc-provider/blob/d7a77119df71fa9eb82cb8282377d23a9c5a0555/lib/models/formats/dynamic.js#L6
- https://github.com/panva/node-oidc-provider/blob/d7a77119df71fa9eb82cb8282377d23a9c5a0555/lib/models/base_model.js#L42

There are several articles on the internet telling, one should NOT use `constructor.name`, because of mentioned reason.
Please let me know if you have any fix or workaround for this issue.

Thank you very much.
